x-pack/plugins/enterprise_search/public/applications/app_search/components/result_settings/logic/result_settings_logic.ts

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { kea, MakeLogicType } from 'kea';
+
+import { Schema, SchemaConflicts } from '../../../../shared/types';
+
+import {
+  FieldResultSetting,
+  FieldResultSettingObject,
+  OpenModal,
+  ServerFieldResultSettingObject,
+} from '../types';
+
+import {
+  areFieldsAtDefaultSettings,
+  clearAllFields,
+  clearAllServerFields,
+  convertServerResultFieldsToResultFields,
+  convertToServerFieldResultSetting,
+  resetAllFields,
+  resetAllServerFields,
+  splitResultFields,
+} from './helpers';
+
+interface ResultSettingsActions {
+  openConfirmResetModal(): void;
+  openConfirmSaveModal(): void;
+  closeModals(): void;
+  initializeResultFields(
+    serverResultFields: ServerFieldResultSettingObject,
+    schema: Schema,
+    schemaConflicts?: SchemaConflicts
+  ): {
+    serverResultFields: ServerFieldResultSettingObject;
+    resultFields: FieldResultSettingObject;
+    schema: Schema;
+    schemaConflicts: SchemaConflicts;
+    nonTextResultFields: FieldResultSettingObject;
+    textResultFields: FieldResultSettingObject;
+  };
+  clearAllFields(): void;
+  resetAllFields(): void;
+  updateField(
+    fieldName: string,
+    settings: FieldResultSetting
+  ): { fieldName: string; settings: FieldResultSetting };
+  saving(): void;
+}
+
+interface ResultSettingsValues {
+  dataLoading: boolean;
+  saving: boolean;
+  openModal: OpenModal;
+  nonTextResultFields: FieldResultSettingObject;
+  textResultFields: FieldResultSettingObject;
+  resultFields: FieldResultSettingObject;
+  serverResultFields: ServerFieldResultSettingObject;
+  lastSavedResultFields: FieldResultSettingObject;
+  schema: Schema;
+  schemaConflicts: SchemaConflicts;
+}
+
+export const ResultSettingsLogic = kea<MakeLogicType<ResultSettingsValues, ResultSettingsActions>>({
+  path: ['enterprise_search', 'app_search', 'result_settings_logic'],
+  actions: () => ({
+    openConfirmResetModal: () => true,
+    openConfirmSaveModal: () => true,
+    closeModals: () => true,
+    initializeResultFields: (serverResultFields, schema, schemaConflicts) => {
+      const resultFields = convertServerResultFieldsToResultFields(serverResultFields, schema);
+      for (const fieldName of Object.keys(schema)) {
+        if (!(fieldName in serverResultFields)) {
+          serverResultFields[fieldName] = {};
+        }
+      }
```

Comment:
map suggestion

```suggestion
      Object.keys(schema).forEach((schemaField) => {
        if (!serverResultFields.hasOwnProperty(schemaField)) {
          serverResultFields[schemaField] = {};
        }
      });
```

Also, can we consider moving this to the utility/helpers file and naming it something like

```js
export const ensureServerFieldsHaveSchema = ...
```

and then here, this action simply just becomes
```js
const resultFields = convertServerResultFieldsToResultFields(serverResultFields, schema);
ensureServerFieldsHaveSchema(serverResultFields);

return {
  // ...
```

Although... wait, am I understanding this correctly? We're only checking for missing schema in `serverResultFields` but not in `resultFields`? Does that seem problematic to you?

Reply:
This action will inject empty or default "fields" into "*ResultFields" objects ... you can see this happening in the specs here: https://github.com/elastic/kibana/pull/94629/files#diff-c97bcbc396488bfc153b85619056a4c5bd5d12602a71ffb299c799d99909f5eaR100

The "baz" property existed in the schema, but did not exist in any existing settings. Therefore, baz is injected into both the `serverResultFields` and `resultFields`.

serverResultFields:
`baz: {},`

resultFields:
`
            baz: {
              raw: false,
              snippet: false,
              snippetFallback: false,
            },
`

So, the code you are pointing out handles the `serverResultFields`. I honestly can't figure out from reading the code where `baz` gets injected into `resultFields`, but it *does* happen somewhere.

Reply:
Your suggestion: 8646d41796cef61efa7abcde8f742ab1df4dbad8

Reply:
Ah I see (re: it happening in `convertServerResultFieldsToResultFields`). Just thinking out loud, that feels a little opaque to me personally - I think functionally something like this makes more sense to me:

```js
serverResultFields = ensureServerFieldsHaveSchema(serverResultFields, schema);
const resultFields = convertServerResultFieldsToResultFields(serverResultFields); // Now you don't have to pass `schema` to this helper, just check for empty objs
```

maybe something to look into later during a refactor - collapsing this for now!